yt_idefix/data_structures.py

```diff
@@ -320,3 +321,108 @@ def _parse_parameter_file(self):
+            unit_system=unit_system,
+            units_override=units_override,
+            geometry=geometry,
+            inifile=inifile,
+        )
+
+    def _parse_parameter_file(self):
+        self._parse_header_file()
+        super()._parse_parameter_file()
+        self._get_time()
+
+    def _parse_header_file(self):
+        """Read some metadata from header file 'definitions.h'."""
+        geom_regexp = re.compile(r"^\s*#define\s+GEOMETRY\s+([A-Z]+)")
+        unit_regexp = re.compile(r"^\s*#define\s+(UNIT_\w+)\s+(\S+)")
+        constexpr = re.compile(r"CONST_\w+")
+
+        # definitions.h is presumed to be along with data file
+        if self._definitions_header is None:
+            self._definitions_header = os.path.join(self.fullpath, "definitions.h")
+
+        if os.path.isfile(self._definitions_header):
+            with open(self._definitions_header) as fh:
+                for line in fh.readlines():
+                    geom_match = re.search(geom_regexp, line)
+                    unit_match = re.search(unit_regexp, line)
+                    if geom_match is None and unit_match is None:
+                        continue
+                    elif unit_match:
+                        unit = unit_match.group(1).lower()
+                        expr = unit_match.group(2)
+                        # The pre-defined constants in definitions should be
+                        # replace with its value
+                        expr = re.sub(constexpr, self._get_constants, expr)
+                        self.parameters[unit] = eval(expr)
+                    elif geom_match:
+                        self.geometry = geom_match.group(1).lower()
+        elif self.geometry is None:
+            raise ValueError(
+                f"Header file {self._definitions_header} couldn't be found. "
+                "The 'geometry' keyword argument must be specified."
+            )
+        else:
+            warnings.warn(
+                f"Header file {self._definitions_header} couldn't be found. "
+                "The code units are set to be 1.0 in cgs by default."
+            )
+
+    def _get_constants(self, matched):
+        """Replace matched constant string with its value"""
+        key = matched.group()
+        return str(pluto_def_constants[key])
+
+    def _get_time(self):
+        """Get current time from vtk.out."""
+        log_file = os.path.join(self.fullpath, "vtk.out")
+        match = re.search(r"\.(\d*)\.", self.parameter_filename)
+        if match is None:
+            raise RuntimeError(
+                f"Failed to parse output number from file name {self.parameter_filename}"
+            )
+        index = int(match.group(1))
+
+        if os.path.isfile(log_file):
+            with open(log_file) as fh:
+                for line in fh.readlines():
+                    if line.startswith(str(index)):
+                        self.current_time = np.float32(line.split()[1])
```

Comment:
A small improvement here: no need to keep reading after we find the value.
Also, I suggest raising an error (or possibly warning ?) in case this doesn't work. I leave this up to up and left a `...` where the error should be.
Note that `else` blocks in for loops read as "if no break".
```suggestion
                        self.current_time = np.float32(line.split()[1])
                        break
                else:
                    ...
```

Reply:
Make sense. It's better.

Reply:
I just found a bug when parse the value. So I'll use regular expression in the next commit to make it more secure. I've tested it locally, though I think it would be better to have a test for this part in the future (no good idea now).

Reply:
Interesting. I didn't suggest using a regexp because I didn't have any idea how to break your implementation. Indeed if you know how to break it, it's worth adding a test. I might be able to make some suggestions how to if you can describe the bug first

Reply:
It's a very wierd case ideed. When some information is removed manually in the log file, for example, the info of first snapshot is missing, `startswith(1)` would locate 10th snapshot info. This case must be ridiculous.

Actually the test I mentioned is for those `else-then-warning` part.

Reply:
Do you think use a regexp here is too expansive? like this:

``` python
 for line in fh.readlines():
     log_match = re.search(log_regexp, line)
     if log_match.group(1) == str(index):
         self.current_time = np.float64(log_match.group(2))
```

Reply:
Oh I see. Well, manually editing a log file sounds like a terrible idea, _but_ you never know, and it's conceivable that a log file would end up corrupted e.g. because of a bug in Pluto.
You could write a very simple test case where the log file would exist but be empty.

Reply:
Damn we overlapped. Yes a regexp is fine but you must also check that you actually get a match and not `None` before you retrieve a group.

Reply:
Just see it. Please ignore this commit. I'll amend it.